### PR TITLE
Move spec resolution into mapper

### PR DESCRIPTION
**Pull request by schrockn**
## Summary & Motivation

We convert back and forth from non-spec code paths and then reconstruct the specs that *would* have been used to construct the `multi_asset`, which is a bit strange.

This moves that logic into the factory machinery.

## How I Tested These Changes

BK

### CODE REVIEW

**Comment by schrockn** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22229?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#22229** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22229?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#22228** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22228?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22227** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22227?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22226** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22226?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22225** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22225?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22224** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22224?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22223** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22223?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22222** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22222?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @schrockn and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_core/definitions/decorators/assets_definition_factory.py [hunk 1]
```diff
@@ -349,6 +354,34 @@ def non_spec_based_create(
...
+                                partition_mapping=in_out_mapper.partition_mappings.get(key),
+                            ),
+                        )
+                        for dep_key in asset_deps.get(output_name, [])
+                    ]
+                else:
+                    deps = input_deps
+
+                resolved_specs.append(asset_out.to_spec(key, deps=deps))
+            return resolved_specs
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/decorators/assets_definition_factory.py:383]: We move this strange-ish callback into a more standard up the stack.

File: python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py [hunk 2]
```diff
@@ -509,21 +509,25 @@ def __call__(self, fn: Callable[..., Any]) -> AssetsDefinition:
...
     )
-    return required_resource_keys - arg_resource_keys
+    if arg_resource_keys:
+        check.invariant(
+            len(explicitly_passed_required_resource_keys) == 0,
+            "Cannot pass resource keys and decorate the function with resources",
+        )
+        return set(resource_defs_bound_to_asset.keys()) - arg_resource_keys
+    else:
+        return explicitly_passed_required_resource_keys | set(resource_defs_bound_to_asset.keys())
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py:530]: I found these names super confusing and the invariant was also unclear, so I spelled this out here. We will be continuing to iteratr on this.

File: python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py [hunk 3]
```diff
@@ -511,21 +511,25 @@ def __call__(self, fn: Callable[..., Any]) -> AssetsDefinition:
         )
 
 
-def compute_required_resource_keys(
-    required_resource_keys: Set[str],
-    resource_defs: Mapping[str, ResourceDefinition],
+def compute_required_resource_keys_for_underlying_op(
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py:514]: Are these changes meant to be in this PR? They seem unrelated to "Move spec resolution into mapper"?

**Review by sryza — CHANGES_REQUESTED** [review]

File: python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py (hunk 3 above)
Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py]: agree. i removed these changes and restacked 👍🏻

**Review by sryza — APPROVED** [review]

**Comment by schrockn** [comment]
### Merge activity

* **Jun 6, 3:55 AM EDT**: @schrockn started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/22229).
